sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnnecessaryImportsCheck.kt

```diff
@@ -141,23 +150,68 @@ class UnnecessaryImportsCheck : AbstractCheck() {
                 }
             }
         }
+
+    private fun getDelegatesImportsFilter(propertyDelegates: Collection<KtPropertyDelegate>, bindingContext: BindingContext) =
+        if (bindingContext == BindingContext.EMPTY) {
+            { imp: KtImportDirective ->
+                imp.importedName?.asString().let { it != "getValue" && it != "setValue" && it != "provideDelegate"}
+            }
+        } else {
+            propertyDelegates.flatMap { propDelegate ->
+                bindingContext.get(BindingContext.DELEGATE_EXPRESSION_TO_PROVIDE_DELEGATE_CALL, propDelegate.expression)
+                    .getResolvedCall(bindingContext)
+                    ?.resultingDescriptor
+                    ?.fqNameOrNull()?.let { listOf(it) }
+                    ?: (propDelegate.parent as? KtProperty)?.let { ktProperty ->
+                        (bindingContext.get(BindingContext.DECLARATION_TO_DESCRIPTOR, ktProperty) as? PropertyDescriptor)?.let { propDescriptor ->
+                            propDescriptor
+                                .accessors
+                                .mapNotNull {
+                                    bindingContext.get(BindingContext.DELEGATED_PROPERTY_RESOLVED_CALL, it)
+                                        ?.resultingDescriptor
+                                        ?.fqNameOrNull()
+                                }
+                        }
+                    } ?: emptyList()
+            }.let { resolvedArrayAccesses ->
+                { imp: KtImportDirective ->
+                    imp.importedFqName !in resolvedArrayAccesses
+                }
+            }
+        }
+
+    private fun getInvokeCallsImportsFilter(calls: Collection<KtCallExpression>, bindingContext: BindingContext) =
+        if (bindingContext == BindingContext.EMPTY) {
+            { imp: KtImportDirective -> imp.importedName?.asString() != "invoke" }
+        } else {
+            { imp: KtImportDirective ->
+                if (imp.importedName?.asString() != "invoke") true
+                else {
+                    imp.importedFqName !in calls.mapNotNull { it.getResolvedCall(bindingContext)?.resultingDescriptor?.fqNameOrNull() }
```

Comment:
Here we execute the `calls.mapNotNull { ... }` for every invocation of the filter, although the result of the `map` is the same every time. WDYT of putting the `calls.mapNotNull { ... }` outside of the lambda in the `else` block, where it can be calculated once and referred to from inside the lambda. The same way it is also done for delegates & arrays above.